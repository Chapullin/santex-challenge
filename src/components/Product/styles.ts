import styled from "styled-components";

export const Item = styled.div`
    display: flex;
`;

export const Price = styled.p`
    text-align: center;
    display: inline-block;
    float: right;
    margin-right: 15px;
`;

export const Button = styled.button`
    margin: auto;
    display: inline-block;
    float: right;
    padding: 10px 30px;
    background-color: red;
    border: 5px solid red;
    border-radius: 30px;
    color: white;
    font-weight: bold;
    &:hover{
        cursor: pointer;
        background-color: palevioletred;
        border-color: palevioletred;
    }
`;
