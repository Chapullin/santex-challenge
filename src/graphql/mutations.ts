import { gql } from "@apollo/client";

export const mutationAddItemToOrder = gql`
  mutation addItemToOrder($productVariantId: ID!, $quantity: Int!) {
    addItemToOrder(productVariantId: $productVariantId, quantity: $quantity) {
      __typename
      ... on Order {
        subTotal
      }
      __typename
      ... on NegativeQuantityError {
        message
      }
      __typename
      ... on InsufficientStockError {
        message
      }
      __typename
      ... on OrderModificationError {
        message
      }
      __typename
      ... on OrderLimitError {
        message
      }
    }
  }
`;
