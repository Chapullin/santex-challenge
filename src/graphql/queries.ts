import { gql } from "@apollo/client";

export const getProductList = gql`
  query products {
    products {
      items {
        id
        name
        description
      }
    }
  }
`;

