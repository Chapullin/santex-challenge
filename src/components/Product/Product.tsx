import { useMutation } from "@apollo/client";
import { useStateWithStorage } from "../../hooks/useStateWithStorage";
import {mutationAddItemToOrder} from "../../graphql/mutations";

// TODO extract into a file with cross types itemsTypes, assetType, variantsType
type itemsTypes = {
    id: string | number;
    description: string;
    name: string;
    featuredAsset: assetType
    variants: variantsType[]
};

type assetType = {
    source: string;
};

type variantsType = {
    price: string
    id: number
}

export const Product = (itemProps: itemsTypes) => {
    const {
        id,
        name,
        description,
        featuredAsset,
        variants,
    } = itemProps;
    const { subTotal, setSubTotal } = useStateWithStorage();

    const [addItemToOrder] = useMutation(mutationAddItemToOrder, {
        onCompleted: (data) => {
            if (data.addItemToOrder.__typename === 'Order') {
                setSubTotal(subTotal + data.addItemToOrder.subTotal);
            }
            // TODO implement a logger
            console.log('Success! Product added to cart.');
        },
        onError: () => {
            // TODO implement a logger
            console.error('Error! Something was wrong adding the product to cart.')
        },
    });

  return (<>
      <div key={id}>
          <h3>{name}</h3>
          <p>AR$ {variants[0].price}</p>
          <p>{description}</p>
          <button
              onClick={()=>addItemToOrder({
                  variables: {
                      productVariantId: variants[0].id,
                      quantity: 1,
                  },
              })}
          >
              Add item to cart
          </button>
          <img src={featuredAsset.source} alt=""/>
      </div>
  </>);
}
