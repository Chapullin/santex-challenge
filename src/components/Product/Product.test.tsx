import { MockedProvider } from "@apollo/client/testing";
import {
    render,
    waitFor,
    screen,
    act,
} from "@testing-library/react";
import { mutationAddItemToOrder } from "../../graphql/mutations";
import { Product } from "./Product";
import React from "react";

const itemProps = {
    id: 1,
    description: "description mock",
    name: "name mock",
    variants: [{ id: 1, price: "1111" }],
    featuredAsset: {
        source: "imgUrl",
    },
};

const productMock = {
    request: {
        query: mutationAddItemToOrder,
        variables: {
            productVariantId: "test-id",
            quantity: 1,
        },
    },
    result: {
        data: {
            addItemToOrder: {
                __typename: "Order",
                subTotal: itemProps.variants[0].price,
            },
        },
    },
};


describe("Product", () => {
    it(
        "should show product, image and price ",
        async () => {
            render(
                <MockedProvider mocks={[productMock]} addTypename={false}>
                    <Product {...itemProps} />
                </MockedProvider>
            );

            await act(() => new Promise((resolve) => setTimeout(resolve, 0)));
            const { getByText } = screen;
            const displayedImage = document.querySelector("img") as HTMLImageElement;
            const { src } = displayedImage;

            await waitFor(() => {
                expect(getByText("description mock")).toBeInTheDocument();
                expect(getByText("AR$ 1111")).toBeInTheDocument();
                expect(getByText("name mock")).toBeInTheDocument();
                expect(src).toContain("imgUrl");
            });

            //TODO if the button shows a feedback/reaction on click,
            // it could be tested here with fireEvent.click()
    });

    it("should match snapshot Header", async () => {
        const { container } = render(
            <MockedProvider mocks={[productMock]} addTypename={false}>
                <Product {...itemProps} />
            </MockedProvider>
        );
        await waitFor(() => {
            expect(container.firstChild).toMatchSnapshot();
        });
    });

});
