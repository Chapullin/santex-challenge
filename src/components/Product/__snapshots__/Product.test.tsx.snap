// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should match snapshot Header 1`] = `
<div
  class="sc-bcXHqe dyuvYz"
>
  <div>
    <h2>
      name mock
    </h2>
    <img
      alt="name mock"
      height="auto"
      src="imgUrl"
      width="100%"
    />
    <p>
      description mock
    </p>
    <button
      class="sc-dkrFOg ieELWA"
    >
      ADD ITEM
    </button>
    <p
      class="sc-gswNZR lbEmXL"
    >
      AR$ 
      1111
    </p>
  </div>
</div>
`;
